### ISSUE TITLE
Is this CViViT implementation causal?

### ISSUE THREAD

**Issue by NSavov** [lucidrains/phenaki-pytorch#42]
Thanks for your work! We are attempting to use your code as a base for implementing an existing paper but we meet a few issues.

One of the key differences of Causal ViViT with its predecessor is the use of causal attention. However, in this implementation we cannot find which part ensures of it, as the attention itself is not causal. We do notice that there is a PEG module with a causal option (by default). We see that there is padding happening but we do not understand the purpose. Can you elaborate what this module does? If it does not ensure causal attention, is there another part that does? 
So far we see proof of our trained models not to be causal, we hope to find if we are doing something incorrectly with the code.

Thanks in advance! And thanks again for the code.

**Comment by NSavov** [comment]
Just wanted to inform that we found some serious bugs, related to reshaping. They were causing, along with other issues, the lack of causality that we noticed. For example, the reshape in the PEG module should be a rearrange instead to allow for also transposing when necessary (such transposes happen when batching spatial and then temporal). After fixing those we were able to obtain a causal and hopefully correct CViViT (currently training). As we see this repo might no longer be maintained, if you are interested about more details, let me know.

**Comment by lucidrains** [comment]
@NSavov hey, thank you for attempting to train this! i would def welcome a PR for the PEG module, if you have found that it violated causality somewhere

it has been a while, so i can't seem to spot the issue

**Comment by lucidrains** [comment]
@NSavov also, i believe another group has already trained cvivit successfully

**Comment by NSavov** [comment]
@lucidrains thanks for the response. We have noticed and used the weights of the mentioned group. However, it shares the same problems. 
The problem is [this](https://github.com/lucidrains/phenaki-pytorch/blob/79d7e149c4ebc1978febff1272d21912b6d22fa3/phenaki_pytorch/attention.py#L70) line in the PEG module. We don''t use the reshape feature now and instead we make sure to rearrange ourselves. We also enable the causal attention and with those 2 changes our causality test passes. Once we get our project running and the time allows, we can prepare a pull request. For now just wanted to point this out.

Thanks again for the code, it is very useful to have something as a base.

**Comment by lucidrains** [comment]
@NSavov you are welcome, and a PR would be great

could you let me know if this change addresses [the underlying issue](https://github.com/lucidrains/phenaki-pytorch/commit/c7713179fdd3a353d464cedaf6206659cc10f3c7#diff-3f396b8874449824189613571724e582f3b72ba12064053dee2c3af2e8111059R296)?

**Comment by lucidrains** [comment]
@NSavov come to think of it, could you point me to the passage in the paper where they said they used causal?

**Comment by NSavov** [comment]
Thanks! We will check it out tomorrow. We apply PEG for spatial and temporal, so it might, but I have to confirm. For enabling causal attention, that is what we do as well.

The causal attention requirement is mentioned in the "Encoder architecture" section on page 4:
![afbeelding](https://github.com/user-attachments/assets/5bd3a0e8-7e18-4c90-9db5-2207ec1880f5)

**Comment by lucidrains** [comment]
@NSavov thank you! yes you are correct then, this is a misstep on my part, my bad

**Comment by lucidrains** [comment]
@NSavov  i can add back a spatial PEG if you find it effective

**Comment by lucidrains** [comment]
think it is resolved

**Comment by sroutray** [comment]
@NSavov @lucidrains I am training with the proposed fix below, but I am still seeing non-causal behavior, Could you let me know what worked for you?
```
        temporal_transformer_kwargs = dict(
            dim = dim,
            dim_head = dim_head,
            heads = heads,
            attn_dropout = attn_dropout,
            ff_dropout = ff_dropout,
            causal = True,
            peg = True,
            peg_causal = True,
        
```

**Comment by NSavov** [comment]
@sroutray It is true that the fix is not enough to make it causal. As far as I remember, the problem came from a reshape in PEG, while it should be transpose + reshape based on if we use it in the temporal or spatial part. Here is how we fixed it in [our code](https://github.com/insait-institute/GenieRedux/blob/ca7351cacaba42bfffc31992bb1b009024d1fa4f/models/components/attention.py#L679). We pass a pattern and we rearrange instead of reshaping. With this fix the model passed our causality tests.
